csrc/fusion_segmenter.cpp

```diff
@@ -3579,11 +3579,65 @@ void SegmentCandidateFinder::resolveScalarsInGroup(SegmentedGroup* group) {
   std::vector<Val*> to_visit;
   std::unordered_set<Val*> visited;
 
+  const auto processTV = [&to_visit](TensorView* tv) {
+    for (auto id : tv->getRootDomain()) {
+      to_visit.push_back(id->getMaybeExpandedExtent());
+    }
+    if (tv->domain()->hasRFactor()) {
+      // traverse from root to rfactor and inspect all Expr attrs and outputs
+      std::vector<Val*> all_vals;
+      for (const auto id_expr : StmtSort::getExprsBetween(
+               tv->fusion(),
+               {tv->getRootDomain().begin(), tv->getRootDomain().end()},
+               {tv->getMaybeRFactorDomain().begin(),
+                tv->getMaybeRFactorDomain().end()})) {
+        all_vals.insert(
+            all_vals.end(), id_expr->inputs().begin(), id_expr->inputs().end());
+        all_vals.insert(
+            all_vals.end(),
+            id_expr->outputs().begin(),
+            id_expr->outputs().end());
+        for (const auto attr : id_expr->attributes()) {
+          if (attr->isVal()) {
+            all_vals.push_back(attr->asVal());
+          }
+        }
+        for (const auto val : all_vals) {
+          if (val->isScalar()) {
+            to_visit.push_back(val);
+          } else if (const auto id = dynamic_cast<IterDomain*>(val)) {
+            to_visit.push_back(id->getMaybeExpandedExtent());
+          }
+        }
+      }
+    }
+  };
+
   // Collect all scalar uses in the group
   for (auto expr : group->exprs()) {
     for (auto input : expr->inputs()) {
       if (input->isScalar()) {
         to_visit.push_back(input);
+      } else if (auto tv = dynamic_cast<TensorView*>(input)) {
+        processTV(tv);
+      }
+    }
+    for (auto attr : expr->attributes()) {
+      auto attr_val = dynamic_cast<Val*>(attr);
+      if (!attr_val) {
+        continue;
+      }
+      if (attr_val->isScalar()) {
+        to_visit.push_back(attr_val);
+      } else if (auto tv = dynamic_cast<TensorView*>(attr_val)) {
+        processTV(tv);
+      }
+    }
+    for (auto output : expr->outputs()) {
+      // We must be able to compute output extents for expression, so here we
+      // ensure the scalars involved are all available to this group
+      if (auto tv = dynamic_cast<TensorView*>(output)) {
```

Comment:
Do we have tests to exercise this case?

Reply:
The `ReshapeToPad` test hits this since the missing scalar appeared only in the root domain of the `pad` output in the second segment.

Reply:
Hmm, then the input tensor of the `pad` op should have the same extent, so shouldn't it use the same scalar?

Reply:
In the unsegmented fusion the input has the same extent as you say. But when it is segmented the inputs have their roots set as the rfactor with new input scalars for the extents.

Reply:
> But when it is segmented the inputs have their roots set as the rfactor with new input scalars for the extents.

Are you referring to the inputs of a fusion segment?

Reply:
Yes. In the test we have a segment with a single pad expression. In that case the expression input is a segment input so it has had its extents replaced.

Reply:
OK, so the pad input tensor becomes a segment input. It originally has the root and rfactor domains as it's the output of the reshape op, but we only keep the rfactor domain for segment inputs, so the input only has the root domain that was originally the rfactor domain. Am I right?

Still not sure why the pad output would be a problem...

Reply:
I think you're right; the pad example does not need output scalars. However, the slice example from https://github.com/NVIDIA/Fuser/issues/656#issuecomment-1701593471 does need this. In that case the `slice` output contains `i5` which is not found in the segment and is exact mapped to the remapped rfactor->root of `T3`. The only way to find it is to traverse root->rfactor in the output of the `SliceOp`. In fact, I think since we have the segment input scalars and since every input to a segment expr is either a segment input or the output of another expr in the segment, it should suffice to look at only attributes and outputs of all expressions in the segment...

Reply:
In #912 we see that even when we do not replace sizes with `ceilDiv` expressions in reshape, we can hit this problem. In that case there is no longer a problem in the second segment, but the first segment is 
```
Inputs:                                                                                                                          
  T0_g[ iS0{i0}, iS1{i2}, iS2{i3}, iS3{i4} ], float                                                                              
Outputs:                                                                                                                         
  T8_g[ iS40{i0}, iS47{( i2 / i5 )}rf, iS48{( (nvfuser_index_t)(i5) )}rf, iS42{i3}, iS43{i4} ], float                            
                                                                                                                                 
%kernel_math {                                                                                                                   
T8_g[ iS40{i0}, iS47{( i2 / i5 )}rf, iS48{( (nvfuser_index_t)(i5) )}rf, iS42{i3}, iS43{i4} ] = view( T0_g[ iS0{i0}, iS1{i2}, iS2{
i3}, iS3{i4} ] )                                                                                                                 
} 
```
and since `i5` is not an input to the `ViewOp`, it is not included as a segment input. We could of course have the new shapes included in `ViewOp::inputs()`instead which would also address this particular case.

Reply:
This makes sense and seems reasonable. Can you please leave a little more detailed comment why this is necessary?